xonsh/imphooks.py

```diff
@@ -0,0 +1,78 @@
+"""Import hooks for importing xonsh source files. This module registers
+the hooks it defines when it is imported.
+"""
+from __future__ import print_function, unicode_literals
+import os
+import sys
+import builtins
+from importlib.machinery import ModuleSpec
+from importlib.abc import MetaPathFinder, SourceLoader
+
+from xonsh.tools import string_types
+from xonsh.execer import Execer
+
+class XonshImportHook(MetaPathFinder, SourceLoader):
+    """Implements the import hook for xonsh source files."""
+
+    def __init__(self, *args, **kwargs):
+        super(XonshImportHook, self).__init__(*args, **kwargs)
+        self._filenames = {}
+        self._execer = None
+
+    @property
+    def execer(self):
+        if hasattr(builtins, '__xonsh_execer__'):
+            execer = builtins.__xonsh_execer__
+            if self._execer is not None:
+                self._execer = None
+        elif self._execer is None:
+            self._execer = execer = Execer(unload=False)
+        else:
+            execer = self._execer
+        return execer
+    #
+    # MetaPathFinder methods
+    #
+    def find_spec(self, fullname, path, target=None):
+        """Finds the spec for a xonsh module if it exists."""
+        spec = None
+        path = sys.path if path is None else path
```

Comment:
> I don't think that this is the sort of decision that xonsh should make for users since it could also affect Python module loading. It is better for folks who want this behaviour to stick it in their xonshrc, I think.

Shouldn't we try to mirror Python's mechanism, though?  I absolutely don't think we want to mess with `sys.path`, but if this import hook is using its own path (what is passed in to `find_spec`?), then I do think it makes sense to add a `'.'` to the start of that (I tried to do something like this non-destructively above).  Without this, it also seems that relative imports without `.` won't work (e.g., a xonsh file `A` should be able to import a file `B` in the same module by doing `import B` _or_ `from . import B`, no?)